Add serialization support for new type (baseclass/subclass relations)

chirags27 opened the pull request:
#### a3c4cf83fb2630f50af6aae5a6e5e67c54e6bf69
<pre>
Add serialization support for new type (baseclass/subclass relations)
rdar://problem/103048679

Reviewed by Alex Christensen.

* Source/WebKit/Scripts/generate-serializers.py:
(SerializedType.__init__):
(SerializedType.subclass_enum_name):
(SerializedType):
(SerializedType.function_name_for_enum):
(MemberVariable.__init__):
(EnumMember.__init__):
(encode_type):
(decode_type.is):
(decode_type):
(generate_impl):
(generate_serialized_type_info):
(parse_serialized_types):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
(IPC::ArgumentCoder&lt;Namespace::Subnamespace::StructName&gt;::decode):
(IPC::ArgumentCoder&lt;Namespace::OtherClass&gt;::decode):
(IPC::ArgumentCoder&lt;Namespace::ReturnRefClass&gt;::decode):
(IPC::ArgumentCoder&lt;WebCore::InheritsFrom&gt;::decode):
(IPC::ArgumentCoder&lt;WebCore::InheritanceGrandchild&gt;::decode):
(IPC::ArgumentCoder&lt;WebCore::TimingFunction&gt;::encode):
(IPC::ArgumentCoder&lt;WebCore::TimingFunction&gt;::decode):
(WTF::void&gt;):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.h:
* Source/WebKit/Scripts/webkit/tests/SerializedTypeInfo.cpp:
* Source/WebKit/Scripts/webkit/tests/TestSerializedType.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;WebCore::TimingFunction&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;WebCore::TimingFunction&gt;::decode): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCAAnimationRemote.mm:
(WebKit::PlatformCAAnimationRemote::Properties::encode const):

Canonical link: <a href="https://commits.webkit.org/257739@main">https://commits.webkit.org/257739@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1fe96628d2ed58c6cdde9cb5a33f585fa5dab061

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99891 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9063 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32972 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109244 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/9893 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/86356 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/92343 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/107151 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105657 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/7471 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/90788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/92343 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/103403 "Passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89379 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/22178 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/92343 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/2868 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/23690 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2820 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/46067 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/8953 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/43163 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2721 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/4676 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

chirags27 commented on Source/WebKit/Scripts/webkit/tests/SerializedTypeInfo.cpp:
```diff
@@ -100,6 +101,8 @@ Vector<SerializedTypeInfo> allSerializedTypes()
             "float"_s,
             "float"_s,
         } },
+        { "Ref<WebCore::TimingFunction>"_s, {
```
@achristensen07 does this look correct?

achristensen07 replied:
No.  This information will need to be exposed to the IPC testing API in a new way that doesn't use allSerializedTypes.  Let's omit that for now.

achristensen07 commented on Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.h:
```diff
@@ -117,6 +118,11 @@ template<> struct ArgumentCoder<WebCore::FloatBoxExtent> {
 };
 
+template<> struct ArgumentCoder<Ref<WebCore::TimingFunction>> {
+    static void encode(Encoder&, const Ref<WebCore::TimingFunction>&);
```
The parameter should be a const WebCore::TimingFunction& not a Ref.

achristensen07 commented on Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
```diff
@@ -515,6 +516,76 @@ std::optional<WebCore::FloatBoxExtent> ArgumentCoder<WebCore::FloatBoxExtent>::d
     };
 }
 
+enum class WebCore_TimingFunction_Subclass : uint8_t {
```
I would expect an isValidEnum function to be generated for this so the decoder works.

achristensen07 commented on Source/WebKit/Scripts/generate-serializers.py:
```diff
@@ -656,10 +728,14 @@ def parse_serialized_types(file, file_name):
         if match:
             attributes, struct_or_class, namespace, name = match.groups()
             continue
+        match = re.search(r'\[(.*)\] (struct|class|alias) (.*)::(.*)\s+(.*) {', line)
```
Let's not allow aliases here.

achristensen07 commented on Source/WebKit/Scripts/generate-serializers.py:
```diff
@@ -59,17 +60,37 @@ def __init__(self, struct_or_class, namespace, name, parent_class_name, members,
+                self.members_are_subclasses = True
+
+    @staticmethod
+    def _wrap_with_ref(name):
```
Unused. Please remove.

achristensen07 commented on Source/WebKit/Scripts/generate-serializers.py:
```diff
@@ -59,17 +60,37 @@ def __init__(self, struct_or_class, namespace, name, parent_class_name, members,
+        return 'Ref<' + name + '>'
 
     def namespace_and_name(self):
+        result = None
```
I don't think this change is necessary.

achristensen07 commented on Source/WebKit/Scripts/generate-serializers.py:
```diff
@@ -59,17 +60,37 @@ def __init__(self, struct_or_class, namespace, name, parent_class_name, members,
     def namespace_unless_wtf_and_name(self):
         if self.namespace == 'WTF':
-            return self.name
+                return self.name
```
Remove excess indentation.

achristensen07 commented on Source/WebKit/Scripts/generate-serializers.py:
```diff
@@ -108,11 +129,13 @@ def is_nested(self):
 
 
 class MemberVariable(object):
```
I don't think we should reuse the MemberVariable class for this.  Let's make a new class.

chirags27 replied:
Yeah, I was thinking of adding a new class, but we'd need similar set of properties to avoid random runtime errors here. Also, it does look like that would cause confusion around what object is stored inside `members = []`.

We could have a new container inside SerializedType and a new class though, but might be overkill. 

Thoughts?

achristensen07 commented on Source/WebKit/Scripts/generate-serializers.py:
```diff
@@ -296,24 +318,41 @@ def encode_type(type):
+        result.append('    std::optional<' + type.subclass_enum_name() + '> type;')
+        result.append('    decoder >> type;')
+        result.append('    if (!type)')
+        result.append('        return std::nullopt;')
```
nit: please add newline after this.

achristensen07 commented on Source/WebKit/Scripts/generate-serializers.py:
```diff
@@ -473,6 +536,29 @@ def generate_impl(serialized_types, serialized_enums, headers):
+        result.append('')
+        if type.condition is not None:
+            result.append('#if ' + type.condition)
+        result.append('template<> bool ' + type.function_name_for_enum() + '<' + type.subclass_enum_name() + ', void>(uint8_t value)')
```
This is mostly duplicate code.  Can you reuse the enum part of generate_impl?

chirags27 replied:
There are some minor differences here from that code, and we'd have to pass around a few parameters to handle that, so I was wondering if this is better...

achristensen07 commented on Source/WebKit/Scripts/generate-serializers.py:
```diff
@@ -533,10 +619,11 @@ def generate_serialized_type_info(serialized_types, serialized_enums, headers):
-        for member in type.members:
-            result.append('            "' + member.type + '"_s,')
-        result.append('        } },')
+        if not type.members_are_subclasses:
```
Nit: I like the less-indented "continue" version of skipping things.

achristensen07 commented on Source/WebKit/Scripts/generate-serializers.py:
```diff
@@ -296,24 +318,41 @@ def encode_type(type):
             result.append('#endif')
+
+    if type.members_are_subclasses:
+        result.append('    ASSERT_NOT_REACHED();')
```
Let's omit this.

achristensen07 commented on Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
```diff
@@ -515,10 +516,93 @@ std::optional<WebCore::FloatBoxExtent> ArgumentCoder<WebCore::FloatBoxExtent>::d
 
 namespace WTF {
 
+template<> bool isValidEnum<WebCore_TimingFunction_Subclass, void>(uint8_t value)
```
Does it compile successfully without this?  Strange if so, but if so then let's omit this.

chirags27 replied:
Yeah, it did look like the build was successful.

chirags27 committed:
Add serialization support for new type (baseclass/subclass relations)
rdar://problem/103048679

Reviewed by Alex Christensen.

* Source/WebKit/Scripts/generate-serializers.py:
(SerializedType.__init__):
(SerializedType.subclass_enum_name):
(SerializedType):
(SerializedType.function_name_for_enum):
(MemberVariable.__init__):
(EnumMember.__init__):
(encode_type):
(decode_type.is):
(decode_type):
(generate_impl):
(generate_serialized_type_info):
(parse_serialized_types):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
(IPC::ArgumentCoder<Namespace::Subnamespace::StructName>::decode):
(IPC::ArgumentCoder<Namespace::OtherClass>::decode):
(IPC::ArgumentCoder<Namespace::ReturnRefClass>::decode):
(IPC::ArgumentCoder<WebCore::InheritsFrom>::decode):
(IPC::ArgumentCoder<WebCore::InheritanceGrandchild>::decode):
(IPC::ArgumentCoder<WebCore::TimingFunction>::encode):
(IPC::ArgumentCoder<WebCore::TimingFunction>::decode):
(WTF::void>):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.h:
* Source/WebKit/Scripts/webkit/tests/SerializedTypeInfo.cpp:
* Source/WebKit/Scripts/webkit/tests/TestSerializedType.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<WebCore::TimingFunction>::encode): Deleted.
(IPC::ArgumentCoder<WebCore::TimingFunction>::decode): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCAAnimationRemote.mm:
(WebKit::PlatformCAAnimationRemote::Properties::encode const):

Canonical link: https://commits.webkit.org/257739@main
